File: grupo1/projC/src/sale.c
#include "sale.h"

#include <string.h>
#include <stdlib.h>

#include <stdio.h>

struct sale
{
    char *  product; /**< Código de produto */
    char *  client; /**< Código de cliente */
    double  price; /**< Preço do produto */
    int     units; /**< Número de unidades vendidas */
    char    promotion; /**< Tipo de promoção N ou P */
    int     month; /**< Número do mês */
    int     branch; /**< Número da filial */
};


Sale isValidSale(char* sale, Clients client_catalogue, Products product_catalogue) {
    Sale new = malloc(sizeof(struct sale));
    char* _temp = strdup(sale);
    char* _fields[100];
    char *_token;
    const char s[2] = " ";
    int i = 0;


    _token = strtok(_temp, s);

    while (_token != NULL) {
        _fields[i] = _token;
        ++i;
        _token = strtok(NULL,s);
    }
    new->product = strdup(_fields[0]);
    new->client  = strdup(_fields[4]);
    new->price      = atof( _fields[1]);
    new->units      = atoi( _fields[2]);
    new->promotion  =       _fields[3][0];
    new->month      = atoi( _fields[5]);
    new->branch     = atoi( _fields[6]);
    free(_temp);
    free(_token);

    if (verifyClient(new->client)                           &&
        verifyProduct(new->product)                         &&
        (new->price <= 999.99    && new->price >= 0)               &&
        (new->units <= 200       && new->units >= 0)                  &&
        (new->promotion == 'N'   || new->promotion == 'P')    &&
        (new->month < 13         && new->month > 0)                   &&
        (new->branch < 4         && new->branch > 0))                 {

        Client _client = newClient(new->client);
        Product _product = newProduct(new->product);

            if ( existClient(client_catalogue, _client) && existProduct(product_catalogue, _product))
            {
                destroyClient(_client);
                destroyProduct(_product);
                return new ;
            }

        destroyClient(_client);
        destroyProduct(_product);
        }
    destroySale(new);
    return NULL;
}

char* getProduct(Sale sale) {
    return sale->product;
}

char* getClient(Sale sale) {
    return sale->client;
}

double getPrice(Sale sale) {
    return sale->price;
}

int getUnits(Sale sale) {
    return sale->units;
}

char getPromotion(Sale sale) {
    return sale->promotion;
}

int getMonth(Sale sale) {
    return sale->month;
}

int getBranch(Sale sale) {
    return sale->branch;
}


void destroySale(Sale sale) {
    free(sale->product);
    free(sale->client);
    free(sale);
}

File: grupo1/projJava/src/model/Sale.java
package model;

import java.io.Serializable;

import static Utils.Constants.N_BRANCHES;

/**
 * Classe auxiliar Sale
 */
public class Sale implements ISale, Serializable {
    private static final long serialVersionUID = 4751558143036898335L;
    private String  product;
    private String  client;
    private double price;
    private int     units;
    private char    promotion;
    private int     month;
    private int     branch;

    /**
     * Construtor com String para a classe
     * @param sale_line Linha com informacao de uma venda
     */
    public Sale(String sale_line) {
        String[] saleGroups = sale_line.split(" ");
        if(saleGroups.length != 7) this.price = -1;
        else {
            this.product = saleGroups[0];
            this.price = Double.parseDouble(saleGroups[1]);
            this.units = Integer.parseInt(saleGroups[2]);
            this.promotion = saleGroups[3].toCharArray()[0];
            this.client = saleGroups[4];
            this.month = Integer.parseInt(saleGroups[5]);
            this.branch = Integer.parseInt(saleGroups[6]);
        }
    }

    /**
     * Verifica se uma venda é válida
     * @return True se a venda for válida e False se for inválida
     */
    public boolean validSale() {
        return this.price >= 0
                && this.price<= 999.99
                && this.units >= 0
                && this.units <= 250
                && month <= 12
                && month >= 1
                && this.branch > 0
                && this.branch <= N_BRANCHES;
    }

    /**
     * Get product
     * @return Devolve o código do produto
     */
    public String getProduct() {
        return product;
    }

    /**
     * Get client
     * @return Devolve o código do cliente
     */
    public String getClient() {
        return client;
    }

    /**
     * Get price
     * @return Devolve o preço
     */
    public double getPrice() {
        return price;
    }

    /**
     * Get units
     * @return Devolve o número de unidades vendidas
     */
    public int getUnits() {
        return units;
    }

    /**
     * Get promotion
     * @return Devolve o tipo de promoção
     */
    public char getPromotion() {
        return promotion;
    }

    /**
     * Get month
     * @return Devolve o mês
     */
    public int getMonth() {
        return month;
    }

    /**
     * Get branch
     * @return Devolve o número da filial
     */
    public int getBranch() {
        return branch;
    }
}